include/podio/ROOTDataSource.h

```diff
@@ -0,0 +1,178 @@
+
+  ///
+  /// @brief Retrieve from podio::ROOTDataSource per-thread readers for the
+  ///        desired columns.
+  ///
+  template <typename T>
+  std::vector<T**> GetColumnReaders(std::string_view columnName);
+
+  ///
+  /// @brief Inform podio::ROOTDataSource that an event-loop is about to
+  ///        start.
+  ///
+  void Initialize() override;
+
+  ///
+  /// @brief Retrieve from podio::ROOTDataSource a set of ranges of entries
+  ///        that can be processed concurrently.
+  ///
+  std::vector<std::pair<ULong64_t, ULong64_t>> GetEntryRanges() override;
+
+  ///
+  /// @brief Inform podio::ROOTDataSource that a certain thread is about to
+  ///        start working on a certain range of entries.
+  ///
+  void InitSlot(unsigned int slot, ULong64_t firstEntry) override;
+
+  ///
+  /// @brief Inform podio::ROOTDataSource that a certain thread is about to
+  ///        start working on a certain entry.
+  ///
+  bool SetEntry(unsigned int slot, ULong64_t entry) override;
+
+  ///
+  /// @brief Inform podio::ROOTDataSource that a certain thread finished
+  ///        working on a certain range of entries.
+  ///
+  void FinalizeSlot(unsigned int slot) override;
+
+  ///
+  /// @brief Inform podio::ROOTDataSource that an event-loop finished.
+  ///
+  void Finalize() override;
+
+  ///
+  /// @brief Returns a reference to the collection of the dataset's column
+  ///        names
+  ///
+  const std::vector<std::string>& GetColumnNames() const override;
+
+  ///
+  /// @brief Checks if the dataset has a certain column.
+  ///
+  bool HasColumn(std::string_view columnName) const override;
+
+  ///
+  /// @brief Type of a column as a string. Required for JITting.
+  ///
+  std::string GetTypeName(std::string_view columnName) const override;
+
+protected:
+  ///
+  /// @brief Type-erased vector of pointers to pointers to column
+  ///        values --- one per slot.
+  ///
+  Record_t GetColumnReadersImpl(std::string_view name, const std::type_info& typeInfo) override;
+
+  std::string AsString() override {
+    return "Podio data source";
+  };
+
+private:
+  /// Number of slots/threads
+  unsigned int m_nSlots = 1;
+
+  /// Input filename
+  std::vector<std::string> m_filePathList = {};
+
+  /// Total number of events
+  unsigned int m_nEvents = 0;
+
+  /// Ranges of events available to be processed
+  std::vector<std::pair<ULong64_t, ULong64_t>> m_rangesAvailable = {};
+
+  /// Ranges of events available ever created
+  std::vector<std::pair<ULong64_t, ULong64_t>> m_rangesAll = {};
+
+  /// Column names
+  std::vector<std::string> m_columnNames{};
+
+  /// Column types
+  std::vector<std::string> m_columnTypes = {};
+
+  /// Collections, m_Collections[columnIndex][slotIndex]
+  std::vector<std::vector<const podio::CollectionBase*>> m_Collections = {};
+
+  /// Active collections
+  std::vector<unsigned int> m_activeCollections = {};
+
+  /// Root podio readers
+  std::vector<std::unique_ptr<podio::ROOTReader>> m_podioReaders = {};
```

Comment:
```suggestion
  std::vector<std::unique_ptr<podio::Reader>> m_podioReaders = {};
```
This should now be possible (with `#include "podio/Reader.h"`)

Reply:
Done